tests/strategies/test_ddp_hydra_support.py

```diff
@@ -0,0 +1,210 @@
+import logging
+import os
+import subprocess
+import sys
+from pathlib import Path
+
+import pytest
+import torch.distributed
+
+from pytorch_lightning import Trainer
+from pytorch_lightning.utilities.imports import _HYDRA_AVAILABLE
+from tests.helpers.boring_model import BoringModel
+from tests.helpers.runif import RunIf
+
+if _HYDRA_AVAILABLE:
+    try:
+        from hydra._internal.callbacks import Callbacks
+    except:
+        Callbacks = None
+
+    from hydra._internal.hydra import Hydra
+    from hydra._internal.utils import create_config_search_path
+    from hydra.types import HydraContext, RunMode
+    from hydra.utils import instantiate
+    from omegaconf import OmegaConf
+
+
+# fixture to run hydra jobs in a clean temporary directory
+# Hydra creates its own output directories and logs
+@pytest.fixture
+def cleandir(tmp_path):
+    """Run function in a temporary directory."""
+    old_dir = os.getcwd()  # get current working directory (cwd)
+    os.chdir(tmp_path)  # change cwd to the temp-directory
+    yield tmp_path  # yields control to the test to be run
+    os.chdir(old_dir)
+    logging.shutdown()
+
+
+# function to run a command line argument
+def run_process(cmd):
+    try:
+        process = subprocess.Popen(
+            args=cmd,
+            shell=False,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        stdout, stderr = process.communicate()
+        if process.returncode != 0:
+            sys.stderr.write(f"Subprocess error:\n{stderr}\n")
+            sys.stderr.write(f"Subprocess stdout:\n{stdout}\n")
+            raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)
+        return stdout, stderr
+    except Exception as e:
+        cmd = " ".join(cmd)
+        sys.stderr.write(f"Error executing:\n{cmd}\n")
+        raise e
+
+
+# Script to run from command line
+script = """
+import hydra
+import torch
+
+from pytorch_lightning import Trainer
+from tests.helpers.boring_model import BoringModel
+
+
+class BoringModelGPU(BoringModel):
+    def on_train_start(self) -> None:
+        # make sure that the model is on GPU when training
+        assert self.device == torch.device(f"cuda:{self.trainer.strategy.local_rank}")
+        self.start_cuda_memory = torch.cuda.memory_allocated()
+
+
+@hydra.main()
+def task_fn(cfg):
+    trainer = Trainer(gpus=cfg.gpus, strategy=cfg.strategy, fast_dev_run=True)
+    model = BoringModelGPU()
+    trainer.fit(model)
+
+
+if __name__ == "__main__":
+    task_fn()
+"""
+
+
+@RunIf(skip_windows=True, min_gpus=2)
+@pytest.mark.skipif(not _HYDRA_AVAILABLE, reason="Hydra not Available")
+@pytest.mark.usefixtures("cleandir")
+@pytest.mark.parametrize("gpus", [1, 2])
```

Comment:
Note that setting `gpus=1` with the DDP strategy will already run distributed. I think we can get away with just testing that (it's faster!)

Reply:
hmm, technically I shouldn't do `gpus=1` because this logic only gets executed for `gpus > 1`.